### PR TITLE
Chronoshifted are force returned to origin with damage by collisions

### LINKED ISSUE #9349: Husks don't chronoshift back to original location
<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26697350-husks-don-t-chronoshift-back-to-original-location?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by atlimit8**
More or less what was wanted on IRC.
Maybe the inflicted damage should be different for the both have `Health` case; I averaged current and max HP of each as rough structural strength based damage values.

~~Fixes #9349~~

### CODE REVIEW

**Comment by penev92** [comment]
TL;DR what does this do? Fixes the "actor X didn't chronoshift back" issue?
If it fixes it reliably then you'd want to reference the issue in the PR description.

File: OpenRA.Mods.RA/Activities/Teleport.cs [hunk 1]
```diff
@@ -53,16 +55,59 @@ public override Activity Tick(Actor self)
...
 
 			Game.Sound.Play(sound, self.CenterPosition);
 			Game.Sound.Play(sound, self.World.Map.CenterOfCell(destination));
 
 			self.Trait<IPositionable>().SetPosition(self, destination);
+			if (force)
+			{
+				if (self.Info.HasTraitInfo<HealthInfo>())
+				{
+					var health = self.Trait<Health>();
```
Comment (abcdefg30) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: Is there a reason why you didn't use `TraitOrDefault` and did a `null`-check above (instead of `HasTraitInfo`)?

**Comment by abcdefg30** [comment]
It works as advertised. :+1: / :white_check_mark: 
Causes stuff like a tank and a husk being on the same cell, but I guess this is a good tradeoff.

(PS: I don't really like the code duplication, but we can't really do something about that, can we?)

**Comment by deniz1a** [comment]
This just destroys tanks when they return from chrono, even when there is nothing on the return cell.

File: OpenRA.Mods.RA/Activities/Teleport.cs [hunk 2]
```diff
@@ -53,16 +55,59 @@ public override Activity Tick(Actor self)
...
+					var givenDamage = (health.HP + health.MaxHP) / 2;
+					foreach (var a in self.World.ActorMap.GetUnitsAt(destination))
```
Comment (pchote) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: The filtering here isn't sufficient. There are certain actor types (e.g. bridges, building bibs) that they should be able to return on without damage).  Use IPositionable.CanEnterCell (might be on a different interface) to check.

**Comment by deniz1a** [comment]
Here's what happens when I chrono a tank:
https://drive.google.com/file/d/0B_dMvsHPqPZkdzIzZnA0MUhYbkE/view

**Comment by atlimit8** [comment]
Works correctly on bibs and bridges - only actual collisions cause damage now.
I accidentally forgot to filter out itself when checking for collisions.

**Comment by atlimit8** [comment]
Merged `foreach` loops.

File: OpenRA.Mods.RA/Activities/Teleport.cs [hunk 3]
```diff
@@ -53,16 +55,71 @@ public override Activity Tick(Actor self)
...
+				destination = bestCell.Value;
+			}
 
 			Game.Sound.Play(sound, self.CenterPosition);
 			Game.Sound.Play(sound, self.World.Map.CenterOfCell(destination));
 
-			self.Trait<IPositionable>().SetPosition(self, destination);
+			var pos = self.Trait<IPositionable>();
+			if (pos.CanEnterCell(destination))
+				force = false;
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: Newline here please.

File: OpenRA.Mods.RA/Activities/Teleport.cs [hunk 4]
```diff
@@ -53,16 +55,71 @@ public override Activity Tick(Actor self)
...
+			pos.SetPosition(self, destination);
+			if (force)
+			{
+				var subcell = pos.OccupiedCells().First(p => p.First == destination).Second;
+				var health = self.TraitOrDefault<Health>();
+				var givenDamage = health != null ? (health.HP + health.MaxHP) / 2 : 0;
+
+				foreach (var a in self.World.ActorMap.GetActorsAt(destination))
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: Name this `actor` please.
There is no need to use a single-character variable name.

File: OpenRA.Mods.RA/Activities/Teleport.cs [hunk 5]
```diff
@@ -53,16 +55,71 @@ public override Activity Tick(Actor self)
...
+
+					var otherHealth = a.TraitOrDefault<Health>();
+					if (otherHealth == null)
+						continue;
+
+					if (health != null)
+					{
+						var maxHealthRadius = health.Info.Radius;
+						if (otherHealth.Info.Radius > maxHealthRadius)
+							maxHealthRadius = otherHealth.Info.Radius;
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: Newline.

File: OpenRA.Mods.RA/Activities/Teleport.cs [hunk 6]
```diff
@@ -53,16 +55,71 @@ public override Activity Tick(Actor self)
...
+						if ((a.CenterPosition - self.CenterPosition).LengthSquared > maxHealthRadius.LengthSquared)
+							continue;
+
+						var otherGivenDamage = (otherHealth.HP + otherHealth.MaxHP) / 2;
+						otherHealth.InflictDamage(a, self, givenDamage, null, true);
+						health.InflictDamage(self, a, otherGivenDamage, null, true);
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: Shouldn't the attacker be the `other` (whether that be self or `actor`)?

Comment (atlimit8) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: They are colliding with each other causing each other damage.

Comment (Phrohdoh) [inline OpenRA.Mods.RA/Activities/Teleport.cs]: Sorry, I read the arguments in the wrong order.

**Comment by atlimit8** [comment]
Changed.

**Comment by abcdefg30** [comment]
Something is wrong now, returning actors don't necessarily cause damage (on buildings):
chronoinweap

I also noticed two other things (not sure if we want to fix them [in this PR]):
- Returning tanks damage a repair pad, although they can move on it
- If a unit is has the Iron Curtain invulnerability, it still gets killed

**Comment by pchote** [comment]
As part of a cleanup of the PR queue I am closing all prs that have had changes requested and no activity for over two weeks. You are welcome to reopen this after you are able to address the review comments above.
